baseapp/baseapp.go

```diff
@@ -711,8 +711,11 @@ func (app *BaseApp) cacheTxContext(ctx sdk.Context, txBytes []byte) (sdk.Context
 // returned if the tx does not run out of gas and if all the messages are valid
 // and execute successfully. An error is returned otherwise.
 func (app *BaseApp) runTx(ctx sdk.Context, mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, result *sdk.Result, anteEvents []abci.Event, priority int64, err error) {
-	// Reset events after each runTx as its not used after the tx is processed
-	defer ctx.MultiStore().ResetEvents()
+	// Reset events after each checkTx or simulateTx or recheckTx
+	// DeliverTx is garbage collected after FinalizeBlocker
+	if mode != runTxModeDeliver {
+		defer ctx.MultiStore().ResetEvents()
+	}
```

Comment:
So the issue was that the deliverTx calls shared the same parent store, and MultiStore.ResetEvents() calls it on the parent store. 

So therefore when goroutine A is clearing events, it can also impact any other goroutine trying to emit events on the child store since its branched later

Reply:
The events aren't needed for simulate, checktx and rechecktx so we can clear it immediate after each runTx